### PR TITLE
Add element-wise operations for arrays

### PR DESCRIPTION
**Pull request by QuentinPepe**
#### Context and Objective

This pull request proposes adding element-wise operations inspired by Julia for arrays in Typst (issue #4106) . The added syntax allows for concise and intuitive element-wise addition, subtraction, multiplication, and division. These changes aim to enhance Typst's numerical computation capabilities.

#### Changes Made

1. **Addition of Element-wise Operators:**
   - Added element-wise operators (`DotAdd`, `DotSub`, `DotMul`, `DotDiv`) to `BinOp`.
   - Updated the logic to handle these new operators in `highlight.rs`, `lexer.rs`, and `parser.rs`.
   - Defined element-wise operations (`elementwise_add`, `elementwise_sub`, `elementwise_mul`, `elementwise_div`) for arrays in `ops.rs`.

2. **Updates to Foundations:**
   - Added the `values` method to access the internal values of arrays in `array.rs`.

3. **Refactoring `SyntaxSet`:**
   - Refactored `SyntaxSet` due to the initial limitation where the number of `SyntaxKind` variants exceeded 128, causing assertions. 
   - Used an array of `u128` to represent the bits, allowing for handling up to 256 variants of `SyntaxKind`.
   - Updated methods to manipulate this new `SyntaxSet`:
     - `new()`
     - `add()`
     - `union()`
     - `contains()`

4. **Unit Tests:**
   - Added comprehensive unit tests in `array.typ` to verify the new element-wise operation functionalities:
     - Basic operations: Addition, subtraction, multiplication, and division.
     - Handling edge cases: Arrays of different lengths, division by zero, mixed data types, empty arrays.
     - Combined and nested operations.

#### Use Case

The new syntax allows for concise and intuitive element-wise operations on arrays:

```typst
let a = (1,2,3)
let b = (4,5,6)

// Element-wise operations
let c = a .+ b
let d = a .- b
let e = a .* b
let f = a ./ b

### CODE REVIEW

**Comment by PgBiel** [comment]
How is this different from https://github.com/typst/typst/pull/3863 ?

**Comment by QuentinPepe** [comment]
> How is this different from #3863 (functionality-wise)?

Thank you for your feedback on my proposal. I understand the concerns raised regarding the addition of new features to the standard library. Here are some key points to explain how my proposal differs and addresses the initial concerns.

1. **Conciseness and Intuitiveness:**
   - **My proposal** uses operators (`.+`, `.-`, `.*`, `./`) that are more intuitive and readable, closely resembling standard mathematical notation. This makes the code easier to write and understand, especially for users coming from languages with similar syntax (like Julia or Matlab).
   - **PR #3863** uses methods (`vector-add`, `vector-sub`, etc.) that are more verbose and may feel less natural for elementary operations.

2. **Seamless Integration:**
   - **My proposal** allows the operators to blend naturally into Typst's existing syntax, facilitating their adoption and use. The operators better align with the language's philosophy of simplicity and readability.
   - **PR #3863** introduces specific new methods, which can complicate the user interface and make learning and using the new features more challenging.

3. **Addressing Initial Concerns:**
   - **Simplicity and Maintainability of the Standard Library:** Elementary operations are basic functionalities widely used in various contexts. Integrating them directly into the standard library simplifies the user experience by avoiding the need for external packages for common operations.

4. **Extensibility:**
   - The proposed operators can also extend to other data structures, offering greater flexibility and broader potential usage. For example, this would allow elementary operations to be applied to matrices or other collections in a natural and consistent manner.

In conclusion, integrating elementary operations directly into Typst with operator syntax improves the language's usability and simplicity. It addresses a general need while providing a more intuitive and widely applicable solution for the Typst user community.

**Comment by johannes-wolf** [comment]
Do these operators also work with scalars? I guess that would be useful, and I could not find it in the tests/code.

**Comment by QuentinPepe** [comment]
> How is this different from #3863 (functionality-wise)?

No, these operators do not currently work with scalars. However, that's a very good idea, and I will add this feature.

File: crates/typst-syntax/src/ast.rs [hunk 1]
```diff
@@ -1431,6 +1431,14 @@ pub enum BinOp {
     MulAssign,
     /// The divide-assign operator: `/=`.
     DivAssign,
+    /// The element-wise addition operator: `.`.
```
Comment (Saecki) [inline crates/typst-syntax/src/ast.rs:1434]: seems like the comment is missing `+`

**Comment by PgBiel** [comment]
> > How is this different from #3863 (functionality-wise)?
> 
> Thank you for your feedback on my proposal. I understand the concerns raised regarding the addition of new features to the standard library. Here are some key points to explain how my proposal differs and addresses the initial concerns.
> (... snip ...)

Thank you for your elaborate response, but this didn't seem to address my initial concern. If I understood correctly, then, this PR implements the same features as #3863, but with different syntax.

Note that your arguments were addressed here: https://github.com/typst/typst/pull/3863#issuecomment-2042409258

Of course, you're free to add to the discussion with new proposals (and I'd love to get to a better solution here, if there is one), but this PR, by itself, didn't seem to add much to it (at least not more than issue #4106 did, which was to propose a new syntax).
Therefore, I'd say that this PR shouldn't move forward unless consensus shifts towards this approach, considering there were little changes compared to the previously closed PR.

**Comment by QuentinPepe** [comment]
I understand your point of view. However, I want to highlight that this PR added scalar handling and could have adapted to other types, which would have broadened its usability. Unfortunately, I acknowledge your decision. Thank you for the time you dedicated to reviewing my pull request.
